### DISCUSSION TITLE
Why is the login request so slow

### DISCUSSION THREAD

**Discussion by KartoffelToby** [pocketbase/pocketbase#2036] [Q&A]
Hey first of all, thanks for this great Project!

I wonder why the Initial login request Take so long (5-10 Seconds). (For the Admin Panel and Client SDK)

Every other request is super fast

**Comment by michal-kapala** [comment]
I've been using `/admins/auth-with-password` extensively today using [js-sdk](https://github.com/pocketbase/js-sdk) and haven't noticed any slowdowns. Are you sure the requests are not getting throttled outside of the application?

**Comment by ganigeorgiev** [comment]
The login requests could be slower compared to the other requests because of the bcrypt hash generation, but not 5-10 seconds slow. Something else is going on in your environment.

Are you able to reproduce the issue using the live demo https://pocketbase.io/demo/?
Are you able to reproduce the issue locally?
Could you provide more information about your environment?

**Comment by KartoffelToby** [comment]
I see ok, the demo Login is super fast.

Looks like its only on my instance, but i wonder why the other requests are fast.

Its only on Login and register requests.

I use a Debian based docker image and compile pocketbase by myself.

**Comment by KartoffelToby** [comment]
![image](https://user-images.githubusercontent.com/8983030/224484052-c2ec01af-c24d-49b5-bdb5-3517204c58d7.png)

This is the timing for /api/collections/users/auth-with-password

**Reply by ganigeorgiev** [reply]
@KartoffelToby I don't know how to reproduce it.

- Are you testing locally?

- Are you testing on prod environment?

- Is there a reverse-proxy (eg. nginx)?

- Do you have any custom code that could cause the delay, (eg. do you have an auth hook)?

**Reply by KartoffelToby** [reply]
@ganigeorgiev i have setup a demo container, from augustodelg/pocketbase:0.13.4 (to get rid of the extra debugging step of self compile) its the same slow /api/admins/auth-with-password request problem
you can test it on https://pocketbase.t-haber.de/_
user: demo@t-haber.de pw: demodemo123
Its sitting behind a reverse proxy from: lucaslorentz/caddy-docker-proxy

**Reply by ganigeorgiev** [reply]
@KartoffelToby You are ignoring my questions and I'm not really sure how to help you debug it. If you can't reproduce the issue locally, then most likely there is something env related.

I'm not familiar with the Docker images you are using and I can't provide support for them.

As mentioned in https://github.com/pocketbase/pocketbase/discussions/2036#discussioncomment-5272131, the `/auth-with-password` requests are expected to be slower than the other requests because of the bcrypt hash but not 5-10sec slow. We currently use a round factor of 13 which could be a little too much for a low spec hardware and I'll consider lowering it to 12 in the next release, but you should be able to get an auth request under ~1 sec.

**Reply by ganigeorgiev** [reply]
To rule out a reverse proxy or some dns issue, you could try starting PocketBase on the server only with the `--http` flag and try to login directly via `http://your-ip:8090/_` and see if that will make any difference.

**Reply by KartoffelToby** [reply]
@ganigeorgiev sry for the leak of infos.

It runs fast on localhost. So it must be the Reverse Proxy. But i wonder why, because all the Headers looks ok. And other Backend Services hasn't this problem.

**Comment by brvier** [comment]
I' m currently having the exact same problem with a low spec VPS, it can handle many concurrent connection, file uploads, and more. But auth-with-password can take more than 30s (i tested without reverse proxy).

**Reply by ganigeorgiev** [reply]
I'm not sure how to reproduce it. As mentioned above, the `auth-with-password` endpoint is expected to be slow but 30s is a little too much. Even on the free fly.io VM (1vCPU, 256MB RAM) the login takes ~1s (see ["User auth with password" benchmarks result](https://github.com/pocketbase/benchmarks/blob/master/results/fly_free_tier.md#users-auth-with-emailpass---small-concurrency-reqs250-conc10)).
Something else is going on in your case but without reproducible steps and/or details about your setup I don't think I can help.

**Comment by brvier** [comment]
I ve rebuilt with more logging, and I confirm it s the issue is with bcrypt, it s not happening everytime, but it s occurs often the night. 
I suspect vps provider slowing down crypt feature, or maybe using an hardware encryption which is already too much used by other VM.

Also on a old atom n270 at home, I don't have any issue.

It s clearly not a pocketbase issue, and i'm just testifying.

Again, thanks a lot for this greatly well writen project.

**Comment by remisharrock** [comment]
I have the same I don't understand because I have a powerfull dedicated server : (scaleway EM-A115X-SSD | 1x Intel Xeon E3 1220 or equivalent 4C / 1T - 3.1 GHz).

I also use cloudflare tunnels but I think this is not the problem because I tried both with the tunnel and directly with an SSH port forwarding directly to my server. For the tests I just login to the admin page /_/ and it takes 5,5 seconds here it is : 

from cloudflare tunnel :

![image](https://github.com/user-attachments/assets/d719a506-a462-489e-a84c-724de4556be8)

from ssh for forwarded :

![image](https://github.com/user-attachments/assets/e67f2bfd-24ea-4a38-b137-823a0d6a6cb7)

how can I improve that ? I have full access to my dedicated server.

Also I tried the brypt algorithm directly on the server command line and it was very fast, here is the command:

time htpasswd -bnBC 12 "nameOfThePassword" YOUR_PASSWORD
nameOfThePassword:$2y$12$WMoh5t9PtrxSQ9tK8pGFlu29lJ2PnNJHUZbEhEJaE/FSTOh52pl7O


real    0m0.222s
user    0m0.222s
sys     0m0.000s

**Reply by ganigeorgiev** [reply]
@remisharrock I'm sorry but I don't think your issue is PocketBase related, or at least I don't see what could be the reason for that.

If you are not able to reproduce the issue locally and when connecting directly to the PocketBase server (_I don't understand what do you mean by "SSH port forwarding directly to my server."_) then the 5s latency is most likely caused by something related with the proxy/env configuration.

**Reply by ganigeorgiev** [reply]
Oh, I forgot to mention that another reason why the login could be slow in PocketBase v0.23+ is that in the recent releases by default we are sending "auth alert" email for new logins. You can disable them from the the auth collection options:

![email_alert_toggle_screenshot](https://i.imgur.com/gXUkeTr.png)

If disabling the toggle "fixes" the issue then the latency is caused by the slow request to the mail server (it could be because of the location of the mail server, throttled/rate limited mail server requests, etc.)

**Reply by remisharrock** [reply]
I think I just found what is the problem and I think it's because I have this:
![image](https://github.com/user-attachments/assets/e8e6d04b-2b68-49a8-a8fc-9124a910a804)
 and then I went to the mail configuration and sent a test email like this:
 
![image](https://github.com/user-attachments/assets/657b7f7d-9589-4eaa-8f68-6e85a775a831)

and it took 5 seconds to fail ! So I think this comes from here ! I'll let you know. 

This is confusing though because you think it's the auth that takes 5 seconds but it's the error while sending the email that takes 5 seconds and you don't see that in the WARN log of the failed attempt to send the email. 

![image](https://github.com/user-attachments/assets/eb82d435-b933-425c-8f70-ffacd6a36f1e)

I'm trying to fix this to see if it works!

**Reply by remisharrock** [reply]
Yes ! you found the issue and our replies crossed ; indeed it works now ! Thank you so much and LONG LIVE POCKETBASE, I love it!

**Reply by evandromendonca** [reply]
This is the answer, tried here, and disabling the sending of alert emails solves the issue. Many VPS are blocking SMTP ports, and this might be the reason for the slowdown, as it waits until it times out.
Thanks @ganigeorgiev for the great work!

**Reply by ganigeorgiev** [reply]
@grawlich The timeout warning message visible in the logs indicates that the toggle hasn't be disabled.

Note that each auth collection (users, _superusers, etc.) has their own toggle, so make sure that you've disabled the toggle for the `_superusers` collection (it is under the "System" collapsible in the collections list sidebar).

_Edit: it seems that the author deleted their post, but the comment remains valid._
